presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java

```diff
@@ -437,6 +439,18 @@ public LookupSourceSupplier createLookupSourceSupplier(
         return createLookupSourceSupplier(session, joinChannels, hashChannel, filterFunctionFactory, sortChannel, searchFunctionFactories, Optional.empty());
     }
 
+    public PagesSpatialIndex createPagesSpatialIndex(
+            Session session,
+            int geometryChannel,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory,
+            List<Integer> outputChannels)
+    {
+        // TODO probably shouldn't copy to reduce memory and for memory accounting's sake
+        List<List<Block>> channels = ImmutableList.copyOf(this.channels);
+        return new PagesSpatialIndexSupplier(session, valueAddresses, types, outputChannels, channels, geometryChannel, spatialRelationshipTest, filterFunctionFactory).get();
```

Comment:
@findepi I wrote a test using `stateful_sleeping_sum` and it passed without any changes. I see 4 `SpatialJoinOperators` sharing a single `SpatialIndexBuilderOperator`, but all `SpatialJoinOperators` run on the same thread.  

`AbstractTestDistributedQueries#testJoinWithStatefulFilterFunction` has the following comment:

```
// With broadcast join, lineitem would be source-distributed and not executed concurrently.
.setSystemProperty(SystemSessionProperties.DISTRIBUTED_JOIN, "true")
```

I assume the same prevents `SpatialJoinOperators` to run concurrently.

@dain Dain, do you happen to know why with broadcast join probe operators don't run concurrently?

Reply:
@dain @findepi I just realized that for probe operators to run concurrently there needs to be more than one page of input. Let me try to modify the test to have that.

Reply:
@dain @findepi I got the test failing on multi-page input: `java.lang.IllegalStateException: 2 threads concurrently executing a stateful function`. I'll go ahead to make the changes necessary to avoid sharing filterFunction instance.